### ISSUE TITLE
Masked attention

### ISSUE THREAD

**Issue by gahdritz** [getkeops/keops#203]
In the MultiHeadAttention implementation on the attention branch, attention masking is not implemented. Is that because it is difficult/impossible to do using KeOps? If that's not the case, how would one add that functionality?

**Comment by gahdritz** [comment]
Hi Jean,

I've had some trouble adding the latter bias term, because it interacts with both symbolic dimensions I and J (and broadcasts along the batch dimension instead). Is it really possible to index both at once? If so, how do I do that?

**Comment by jeanfeydy** [comment]
Hi @gahdritz ,

Indeed, I hadn’t seen that your bias term was dense with respect to both I and J. If that’s the case, there may not be a good way of speed things up with KeOps: the library is all about avoiding the storage and transfer of “I-by-J” variables. 

A work-around could be to encode your bias term as a symbolic matrix defined using appropriate “I” and “J” variables. For instance, if `B_ij`  is a `(I, J)` bias matrix with rank R, SVD decomposition will allow you to write it down as a product `U @ V.T` where `U` is `(I, R)` and `V` is `(J, R)`.
Then, if `u_i` and `v_j` denote the lines of `U` and `V`, respectively, the dot product `(u_i | v_j)` defines a valid KeOps formula that computes the value of `B_ij` and can be added to the standard formula for the attention logits.
(Note that this would work with any number of batch dimensions or attention heads.)

If the rank R is smaller than 32, this could be very efficient…
But obviously, depending on your application and the properties of your bias matrix, such a strategy may not be tractable: what do you think?
Best regards,
Jean

**Comment by gahdritz** [comment]
Unfortunately, in my use case, the rank is in the 100s or even 1000s. Thanks anyway for the help!
